### PR TITLE
[server] clean up org-only migration code

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description

This PR removes a lot of code that is no longer used with regards to user billing and code pathes for the previous working model where users would act as container for projects, workspaces, etc.

This PR:
- removes unused code and API
- makes organizationId mandatory in Workspace and Project

## Related Issue(s)

Fixes WEB-217
Fixes WEB-470
Fixes WEB-471

## How to test

Since the changes are across the board we will need to carefully review the code as well as carefully test the preview environment.

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - se-server-cleanups</li>
	<li><b>🔗 URL</b> - <a href="https://se-server-cleanups.preview.gitpod-dev.com/workspaces" target="_blank">se-server-cleanups.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-payment
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold

### CODE REVIEW

File: components/server/src/jobs/fix-stripe-job.ts [hunk 1]
```diff
@@ -51,7 +51,7 @@ export class FixStripeJob implements Job {
                 const organizationId = row.organizationId;
                 const stripeCustomerId = row.stripeCustomerId;
                 const newAttributionID = AttributionId.render({ kind: "team", teamId: organizationId });
-                const oldAttributionID = AttributionId.render({ kind: "user", userId: userId });
+                const oldAttributionID = "user:" + userId;
```
Comment (svenefftinge) [inline components/server/src/jobs/fix-stripe-job.ts:54]: will remove this job in a follow up PR

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx [hunk 2]
```diff
@@ -178,7 +178,7 @@ export function CreateWorkspacePage() {
     const [selectAccountError, setSelectAccountError] = useState<SelectAccountPayload | undefined>(undefined);
 
     const createWorkspace = useCallback(
-        async (options?: Omit<GitpodServer.CreateWorkspaceOptions, "contextUrl">) => {
+        async (options?: Omit<Omit<GitpodServer.CreateWorkspaceOptions, "contextUrl">, "organizationId">) => {
```
Comment (easyCZ) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:181]: Should be able to write it as `Omit<GitpodServer, "dispose" | "setClient">`

File: components/gitpod-protocol/src/attribution.ts [hunk 3]
```diff
@@ -22,19 +17,15 @@ export interface TeamAttributionId {
...
     }
 
-    export function create(userOrTeam: User | Team): AttributionId {
-        if (User.is(userOrTeam)) {
-            return { kind: "user", userId: userOrTeam.id };
-        } else {
-            return { kind: "team", teamId: userOrTeam.id };
-        }
+    export function create(organization: Organization): AttributionId {
+        return { kind: "team", teamId: organization.id };
```
Comment (easyCZ) [inline components/gitpod-protocol/src/attribution.ts:25]:
```suggestion
        return AttributionId.createFromOrganizationId(organization.id);
```

File: components/server/src/billing/billing-mode.spec.db.ts [hunk 4]
```diff
@@ -72,17 +62,9 @@ class BillingModeSpec {
...
-        //         id: "stripe-123",
-        //         customer: stripeTeamCustomerId,
-        //         isTeam: true,
-        //     };
-        // }
-
         interface Test {
             name: string;
-            subject: User | Pick<Team, "name">;
+            subject: Pick<Team, "name">;
```
Comment (easyCZ) [inline components/server/src/billing/billing-mode.spec.db.ts:67]: Could probs just change it to `teamName: string`

File: components/server/src/billing/entitlement-service-ubp.ts [hunk 5]
```diff
@@ -70,7 +70,7 @@ export class EntitlementServiceUBP implements EntitlementService {
 
     protected async checkUsageLimitReached(
         user: User,
-        organizationId: string | undefined,
+        organizationId: string,
```
Comment (easyCZ) [inline components/server/src/billing/entitlement-service-ubp.ts:73]: out of scope, but we could start using `uuid` type here (and other places) for even more type safety

File: components/server/src/user/stripe-service.ts [hunk 6]
```diff
@@ -93,27 +92,6 @@ export class StripeService {
         return priceInformation.humanReadableDescription;
     }
 
-    async cancelSubscriptionForUser(userId: string) {
```
Comment (easyCZ) [inline components/server/src/user/stripe-service.ts:96]: How is this now handled? I may have missed it in the large diff

File: components/server/src/workspace/gitpod-server-impl.ts [hunk 7]
```diff
@@ -2877,6 +2825,10 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
         }
 
         const team = await this.teamDB.createTeam(user.id, name);
+        // create a cost center
+        await this.usageService.getCostCenter({
```
Comment (easyCZ) [inline components/server/src/workspace/gitpod-server-impl.ts:2829]: What's the reason for this addition?

File: components/server/src/workspace/gitpod-server-impl.ts [hunk 8]
```diff
@@ -4361,20 +4272,11 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
...
-                    "Cannot create Stripe customer profile for another user",
-                );
-            }
-        }
+        const team = (await this.guardTeamOperation(attrId.teamId, "update", "not_implemented")).team;
 
+        //TODO billing email should be editable within the org
         const billingEmail = User.getPrimaryEmail(user);
-        const billingName = attrId.kind === "team" ? team!.name : User.getName(user);
+        const billingName = team!.name;
```
Comment (easyCZ) [inline components/server/src/workspace/gitpod-server-impl.ts:4279]: is the `!` needed here? And if so, do we need to handle when `team` is undefined?

File: components/server/src/workspace/workspace-factory.ts [hunk 9]
```diff
@@ -370,7 +366,7 @@ export class WorkspaceFactory {
     protected async createForCommit(
         ctx: TraceContext,
         user: User,
-        organizationId: string | undefined,
+        organizationId: string,
```
Comment (easyCZ) [inline components/server/src/workspace/workspace-factory.ts:369]: Loving the more concrete types

File: components/server/src/workspace/workspace-starter.ts [hunk 10]
```diff
@@ -921,16 +921,8 @@ export class WorkspaceStarter {
...
-            if (!usageAttributionId) {
-                if (!user.additionalData?.isMigratedToTeamOnlyAttribution) {
-                    usageAttributionId = await this.userService.getWorkspaceUsageAttributionId(user);
-                } else {
-                    if (usageAttributionId === undefined) {
-                        throw new Error("No usage attribution ID found");
-                    }
-                }
+            if (!workspace.organizationId) {
+                throw new Error("No usage attribution ID found");
```
Comment (easyCZ) [inline components/server/src/workspace/workspace-starter.ts:925]:
```suggestion
                throw new Error("No usage attribution ID found - could not derive attribution from workspace organization");
```

Comment (easyCZ) [reply components/server/src/workspace/workspace-starter.ts:925]: This in general looks just like a safety. What are the reasons we'd not have an Org ID on a workspace? Is this older data?

**Review by easyCZ — APPROVED** [review]
LGTM broadly. Will do a preview pass as well

**Comment by easyCZ** [comment]
Looks good from my testing, couldn't find a way to break it

File: components/gitpod-protocol/src/attribution.ts (hunk 3 above)
Comment (geropl) [reply components/gitpod-protocol/src/attribution.ts:25]: Is it really true that nowhere in the system we still have UserAttributionId and see value in having common parsing code for it?
I thought we'd go for removing the constructors/helpers, but keep the underlying data type... but maybe we don't access that data anymore/it's gone?

File: components/server/src/user/stripe-service.ts (hunk 6 above)
Comment (svenefftinge) [reply components/server/src/user/stripe-service.ts:96]: there are no subscriptions for users

File: components/server/src/workspace/gitpod-server-impl.ts (hunk 7 above)
Comment (svenefftinge) [reply components/server/src/workspace/gitpod-server-impl.ts:2829]: To create the cost center immediately and not lazily which is harder to reason about

File: components/server/src/user/stripe-service.ts (hunk 6 above)
Comment (easyCZ) [reply components/server/src/user/stripe-service.ts:96]: Sorry, missed the fact this method is "user" specific.. Makes sense

File: components/gitpod-protocol/src/attribution.ts (hunk 3 above)
Comment (svenefftinge) [reply components/gitpod-protocol/src/attribution.ts:25]: Yes, we no longer have user: attribution and if so it is a bug. I have checked the DB and am only aware of the usage table having some credit entries, which I will fix manually.

**Comment by geropl** [comment]
Starting a second round here... :eyes:

File: components/gitpod-db/src/typeorm/entity/db-project.ts [hunk 11]
```diff
@@ -30,7 +30,7 @@ export class DBProject {
         transformer: Transformer.MAP_EMPTY_STR_TO_UNDEFINED,
     })
     @Index("ind_teamId")
-    teamId?: string;
+    teamId: string;
```
Comment (geropl) [inline components/gitpod-db/src/typeorm/entity/db-project.ts:33]: Here I found:
- 4 projects that are not deleted and don't have this set (SELECT * FROM d_b_project WHERE teamId = '' AND markedDeleted = FALSE;)
- 15k markedDeleted

Would be awesome if we could cleanup that state somehow, so we don't stumble over it in some month's time. :lotus_position:

File: components/server/src/jobs/org-only-migration-job.ts [hunk 12]
```diff
@@ -1,57 +0,0 @@
-/**
```
Comment (geropl) [inline components/server/src/jobs/org-only-migration-job.ts:1]: What is the lower bound of dedicated instances that we can't update from if we remove this job now? Does that affect any of Team SIDs plans regarding updating old instances (I think they want to re-create anyway, just want to make sure at this point).

Comment (svenefftinge) [reply components/server/src/jobs/org-only-migration-job.ts:1]: I'm pretty sure we don't have cells that old that we were creating users as billing cost centers

File: components/gitpod-db/src/typeorm/entity/db-project.ts (hunk 11 above)
Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:33]: will look into this today

File: components/server/src/workspace/gitpod-server-impl.ts (hunk 8 above)
Comment (svenefftinge) [reply components/server/src/workspace/gitpod-server-impl.ts:4272]: it is not needed

File: components/gitpod-db/src/typeorm/entity/db-project.ts (hunk 11 above)
Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:33]: I have checked those four projects and they were all unused so I marked them as deleted as well.

Comment (geropl) [reply components/gitpod-db/src/typeorm/entity/db-project.ts:33]: Thank you! :pray:

**Pull request merged**